refactor(helpers)!: remove v8 deprecated unique

Shinigami92 opened the pull request:
extracted from #2608

- #2608

referencing #1785

- #1785

should be merged before starting #2335

- #2335

Shinigami92 committed:
refactor(helpers)!: remove v8 deprecated unique

ST-DDT commented:
As mentioned earlier. I don't think we should remove unique from faker as it is a very useful feature when generating test data.
I would like to check with the rest of the team and to a certain extend the community, whether we should remove it or replace it.

EDIT: Reason: #2667 in combination with #2664 will likely invalidate the circumstances that the initial decision was made upon.

Please vote this comment with 
- 👍 for removal of unique and 
- 👎 for retaining unique, although with a different implementation and interface.

matthewmayer commented:
What about the possibility of taking an external replacement like https://www.npmjs.com/package/enforce-unique and moving it within the faker-js org, but as a separate package/repo?

Shinigami92 commented:
> As mentioned earlier. I don't think we should remove unique from faker as it is a very useful feature when generating test data.
> 
> I would like to check with the rest of the team and to a certain extend the community, whether we should remove it or replace it.
> 
> 
> 
> Please vote this comment with 
> 
> - 👍 for removal of unique and 
> 
> - 👎 for retaining unique, although with a different implementation and interface.

I need to say that I 100% dislike to bring up a long discussion we already had and even written down in our internal notes. As Matthew already wrote: there is another package that does the same if not better.
And for all the reasons: please lock back in history notes. Period.

matthewmayer commented:
Let's just make sure we have very good documentation explaining how to migrate from using the existing method to a third-party package.

matthewmayer committed:
Add documentation on recommended ways to generate unique values

format docs

fix code

simplify code

move the CJS example

remove the unique section from usage.md and reference new page

matthewmayer commented:
i went ahead and added some documentation and suggested alternatives direct on this branch. Feedback welcome.

Shinigami92 committed:
Fix typos

Shinigami92 commented:
Thank you @matthewmayer 
I rebased it, fixed a conflict and fixed some minor typos

LGTM 👍

ST-DDT commented on docs/guide/unique.md:
````diff
@@ -0,0 +1,120 @@
+1. Use `faker.helpers.uniqueArray()` if you want to generate all the values at one time. For example
+
+```ts
+faker.helpers.uniqueArray(faker.internet.email, 1000); // will generate 1000 unique email addresses
````
IMO it is inconsequetial to remove unique but keep uniqueArray.

Shinigami92 replied:
I dont understand?! uniqueArray uses shuffle which uses int which is based on the seed

unique however does not use seed at all and is a plain utility function not bound to faker's seed

ST-DDT replied:
https://github.com/faker-js/faker/blob/7c7f78daa0fef802af916c2f192cc36bee6c88d2/src/modules/helpers/index.ts#L760

Shinigami92 replied:
https://github.com/faker-js/faker/blob/7c7f78daa0fef802af916c2f192cc36bee6c88d2/src/modules/helpers/index.ts#L751

matthewmayer replied:
Difference is that uniqueArray is just unique for the lifetime of that method call. 

Whereas unique was unique for the lifetime of that faker instance.

ST-DDT replied:
@Shinigami92 If my post wasnt clear enough: There are two branches in that method. One for arrays and one for functions.

Yes, and I would like to have a method that I can pass an array of known ones and it generates a single new one.

Shinigami92 replied:
Did you know when you pass `faker.datatype.boolean({ probability: 0 })` it always return `false` and even dont consume a faker seed.
However, unique does not use something from faker internally at all, in any kind of branch.

But TBH, I'm tired of starting a discussion of something we already discussed to its exhaustiveness half a year ago.

ST-DDT replied:
> However, unique does not use something from faker internally at all, in any kind of branch.

I'm aware.



> But TBH, I'm tired of starting a discussion of something we already discussed to its exhaustiveness half a year ago.

**Important**: My question is not about `unique`, but `uniqueArray(() => fn())`.

Is `uniqueArray(()=>fn())` a faker-ish method for you?

<details><summary>If yes</summary>
Hypothetical: Is a method using uniqueArray for generating random values first and returning a function to return/pop an elememt from that list faker-ish for you?

````ts
const list = uniqueArray(source);
return () => if list.empty throw else yield list.shift();
````

Not that I think about adding/replacing it like that.

</details>

<details>
<summary>If no,</summary>
 we should remove that as well for consistency.
</details>

Shinigami92 replied:
I'm not sure right now, maybe discuss in a meeting 🤔 
I never used `uniqueArray` and I also was not aware that it was already there since the faker-takeover 👀 

But I would definitely see this in another PR for now, cause this PR currently aims for removing the one method that bleeds side-effects and so holds other PR back.

ST-DDT commented on docs/guide/usage.md:
```diff
@@ -233,45 +233,6 @@ Here, we could also pass in the `sex` value as argument, but in our use-case the
-Please have a look at the issue [#1785](https://github.com/faker-js/faker/issues/1785).  
-We will update these docs once a replacement is available.
-:::
+Unlike the `_id` property that uses an `uuid` implementation, which is unique by design, the `email` may contain duplicates. We have a dedicated guide page on generating [unique values](unique).
```
Uuid is not unique by design but by chance.

matthewmayer replied:
Maybe just remove the comment about UUIDs completely? It's not very important for this section of the docs.

ST-DDT replied:
```suggestion
Unlike the `_id` property that uses an `uuid` implementation, which has a low chance of duplicates, the `email` function is more likely to produce duplicates, especially if the call arguments are similar. We have a dedicated guide page on generating [unique values](unique).
```

ST-DDT commented:
Please dont rebase unless necessary as it makes reviews harder.

matthewmayer commented:
This will also need a migration guide snippet. Hopefully it can be quite short now and just link to the unique guide page.

Shinigami92 committed:
Add migration file

Shinigami92 commented:
> This will also need a migration guide snippet. Hopefully it can be quite short now and just link to the unique guide page.

added

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

ST-DDT commented on docs/guide/unique.md:
````diff
@@ -0,0 +1,120 @@
+}
+```
+
+## Migrating from faker.helpers.unique
+
+Prior to v9, Faker provided a [`faker.helpers.unique()`](https://v8.fakerjs.dev/api/helpers.html#unique) method which had a global store to keep track of duplicates. This was removed in v9.
+
+In general you should switch to using one of the strategies above. For example, migrating to `enforce-unique`:
````
IMO this section better suited in the migration/upgrading guide.

Shinigami92 replied:
@matthewmayer could you please take this part over?

matthewmayer replied:
Sure

ST-DDT commented on docs/guide/unique.md:
````diff
@@ -0,0 +1,120 @@
+Basic example:
+
+```js
+//OLD
````
I think most of our code uses a space after the comment slashes e.g. `// OLD`.
Should we change these for consistency or is that just my personal preference?

Shinigami92 replied:
I observed these comments written almost and only by @matthewmayer  
After a while I decided to not care anymore
Either @matthewmayer should use `// ...` or stay at `//...` :shrug:

ST-DDT commented on docs/guide/unique.md:
````diff
@@ -0,0 +1,120 @@
+});
+```
+
+Note `enforce-unique` does not support the `exclude` or `store` options. If you were previously using these, you may wish to build your own unique logic instead.
````
```suggestion
Note: `enforce-unique` does not support the `exclude` or `store` options. If you were previously using these, you may wish to build your own unique logic instead.
```

Or maybe use an actual `::: tip Note` block?

Shinigami92 replied:
I'm for the `::: tip Note` variant, cause that is mostly the one suggested for inside VitePress

Shinigami92 committed:
Update docs/guide/usage.md

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

matthewmayer committed:
move migration info to migration guide

ST-DDT commented on docs/guide/unique.md:
````diff
@@ -0,0 +1,58 @@
+faker.animal.type(); //'horse'
+```
+
+Some methods and locales use a much larger data sets than others. For example `faker.animal.type` has only 13 possible animals to choose from. By contrast `faker.person.fullName()` pulls from a list of hundreds of first names, surnames and prefixes/suffixes, so it can generate hundreds of thousands of unique names. Even then, the [birthday paradox](https://en.wikipedia.org/wiki/Birthday_Paradox)  ...
````
> much larger data sets than others. For example `faker.animal.type` has only 13

Optional: IMO it is strange to talk about larger sets and then talk about having only 13. Maybe flip the comparison around?

matthewmayer replied:
Or change "larger" to "smaller"?

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

matthewmayer committed:
Update unique.md

Larger to smaller

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

Shinigami92 committed:
Apply review suggestions

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

MansurAliKoroglu commented:
I published a new version of `enforce-unique`. Now it supports exclude.

ejcheng commented on docs/guide/unique.md:
````diff
@@ -0,0 +1,58 @@
+faker.animal.type(); //'horse'
+```
+
+Some methods and locales use a much smaller data sets than others. For example `faker.animal.type` has only 13 possible animals to choose from. By contrast `faker.person.fullName()` pulls from a list of hundreds of first names, surnames and prefixes/suffixes, so it can generate hundreds of thousands of unique names. Even then, the [birthday paradox](https://en.wikipedia.org/wiki/Birthday_Paradox) ...
````
```suggestion
Some methods and locales use much smaller data sets than others. For example, `faker.animal.type` has only 13 possible animals to choose from. In contrast, `faker.person.fullName()` pulls from a list of hundreds of first names, surnames, and prefixes/suffixes, so it can generate hundreds of thousands of unique names. Even then, the [birthday paradox](https://en.wikipedia.org/wiki/Birthday_Paradox) means that duplicate values will quickly be generated.
```

ejcheng commented on docs/guide/unique.md:
```diff
@@ -0,0 +1,58 @@
+
+Some methods and locales use a much smaller data sets than others. For example `faker.animal.type` has only 13 possible animals to choose from. By contrast `faker.person.fullName()` pulls from a list of hundreds of first names, surnames and prefixes/suffixes, so it can generate hundreds of thousands of unique names. Even then, the [birthday paradox](https://en.wikipedia.org/wiki/Birthday_Paradox) ...
+
+Sometimes you want to generate unique values, for example you may wish to have unique values in a database email column.  
```
```suggestion
Sometimes, you want to generate unique values. For example, you may wish to have unique values in a database email column.  
```

ejcheng commented on docs/guide/unique.md:
```diff
@@ -0,0 +1,58 @@
+Sometimes you want to generate unique values, for example you may wish to have unique values in a database email column.  
+There are a few possible strategies for this:
+
+1. Use `faker.helpers.uniqueArray()` if you want to generate all the values at one time. For example
```
```suggestion
1. Use `faker.helpers.uniqueArray()` if you want to generate all the values at one time. For example:
```

Shinigami92 committed:
Apply suggestions from code review

Co-authored-by: Eric Cheng <ericcheng9316@gmail.com>

Shinigami92 committed:
Merge branch 'next' into remove-deprecated-unique

ST-DDT reviewed (approved):
Approved as per old team decision so it's no longer hard blocked, but I still think it's wrong.

Shinigami92 commented:
> Approved as per old team decision so it's no longer hard blocked, but I still think it's wrong.

I owe you a favor 👍
